text/3628-async-gen-types.md

```diff
@@ -0,0 +1,154 @@
+- Feature Name: `async_gen_types`
+- Start Date: 2024-05-06
+- RFC PR: [rust-lang/rfcs#3628](https://github.com/rust-lang/rfcs/pull/3628)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Allow the syntax `async T` and `gen T` as types, equivalent to
+`impl Future<Output = T>` and `impl Iterator<Item = T>` respectively. Accept
```

Comment:
Is there particular motivation for desugaring to `impl` rather than desugaring to the trait and allowing the user to specify `impl` vs. `dyn`? The current approach would preclude users doing things like:

```rust
fn takes_iterator(iter: &mut dyn gen T) { ... }`
fn returns_future() -> Pin<Box<dyn async T>> { ... }
```

Allowing `dyn` would make this syntax usable in more places and would help mitigate the existing "cliff" users fall off of when transitioning to dynamic dispatch.

Reply:
That is, I'm imagining that `gen T` would be sugar for `Iterator<Item = T>`, and `async T` would be sugar for `Future<Item = T>`.